[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42944 [c10d][jit] initial torchbind bindings for ProcessGroup
* #47368 [WIP] use lambda to register custom class init
* #47343 [c10d] switch ProcessGroup to be managed by intrusive_ptr
* #47075 [c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr
* #47074 [c10d] switch Store to be managed by intrusive_ptr
* **#44046 [c10d] switch ProcessGroup::Work to be managed by intrusive_ptr**



Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

wanchaol committed:
[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -119,33 +119,33 @@ class ProcessGroup {
   }
 
-  virtual std::shared_ptr<ProcessGroup::Work> broadcast(
+  virtual c10::intrusive_ptr<ProcessGroup::Work> broadcast(
```
This change is probably going to break other third party backends like: https://github.com/intel/torch-ccl/blob/master/src/ProcessGroupCCL.hpp#L136 and https://github.com/openucx/torch-ucc/blob/master/include/torch_ucc.hpp#L77.

I'm guessing this is necessary for TorchScript and there is no way around it, so should we ask the third-party libraries to make this change as well? (we can probably file issues on those repos).

cc @agolynski Since this affects the c10d extension.

wanchaol replied:
Yeah agreed that we should ask them to make the changes. Do you have a list of third party backends or are these two the only two that's currently using c10d extension?

mrshenli replied:
The ones I am aware of are Intel and UCX. See:

1. https://github.com/openucx/torch-ucc
2. https://github.com/pytorch/pytorch/pull/28068

Could you please check with @agolynski, he might know more context.

agolynski replied:
Hello @chengjunlu @mshiryaev @Sergei-Lebedev @srinivas212 

This PR will break master -> master dependency in torch-ccl and torch-ucc.

In near future we'll be changing ProcessGroup API which will break these repos as well. Would it be okay if you depend on 1.6 (upgrade to 1.7 when released) and not on master meanwhile?

@Sergei-Lebedev @srinivas212: How does torch-ucc depend on pytorch, do you require users to install from pytorch from master branch?

srinivas212 replied:
Hi @agolynski @mrshenli @pritamdamania87 - after discussions with both Torch-UCC and Torch-CCL teams, the near term plan is to fix the issue in the third-party repo once this change lands.

In general, it is best we keep third-party plugins in sync w/ master.

wanchaol replied:
going to land the stack soon, created https://github.com/openucx/torch-ucc/pull/23 and https://github.com/intel/torch-ccl/pull/11 to ucc and ccl to do the API migration.

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"


Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup::Work to be managed by intrusive_ptr"




Differential Revision: [D23632280](https://our.internmc.facebook.com/intern/diff/D23632280)

[ghstack-poisoned]

mrshenli commented:
Hey @wanchaol, looks like this PR breaks master, shall we revert?

```
Nov 11 03:04:20 ======================================================================
Nov 11 03:04:20 ERROR [2.020s]: test_all_gather_multigpu_complex (__main__.TestDistBackendWithFork)
Nov 11 03:04:20 ----------------------------------------------------------------------
Nov 11 03:04:20 Traceback (most recent call last):
Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 278, in wrapper
Nov 11 03:04:20     self._join_processes(fn)
Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 395, in _join_processes
Nov 11 03:04:20     self._check_return_codes(elapsed_time)
Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 431, in _check_return_codes
Nov 11 03:04:20     raise RuntimeError(error)
Nov 11 03:04:20 RuntimeError: Processes 0 1 exited with error code 10
Nov 11 03:04:20 
Nov 11 03:04:20 ======================================================================
Nov 11 03:04:20 ERROR [1.920s]: test_reduce_sum_cuda (__main__.TestDistBackendWithFork)
Nov 11 03:04:20 ----------------------------------------------------------------------
Nov 11 03:04:20 Traceback (most recent call last):
Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 278, in wrapper
Nov 11 03:04:20     self._join_processes(fn)
Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 395, in _join_processes
Nov 11 03:04:20     self._check_return_codes(elapsed_time)
Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 431, in _check_return_codes
Nov 11 03:04:20     raise RuntimeError(error)
Nov 11 03:04:20 RuntimeError: Processes 0 1 exited with error code 10
```

wanchaol commented:
> Hey @wanchaol, looks like this PR breaks master, shall we revert?
> 
> ```
> Nov 11 03:04:20 ======================================================================
> Nov 11 03:04:20 ERROR [2.020s]: test_all_gather_multigpu_complex (__main__.TestDistBackendWithFork)
> Nov 11 03:04:20 ----------------------------------------------------------------------
> Nov 11 03:04:20 Traceback (most recent call last):
> Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 278, in wrapper
> Nov 11 03:04:20     self._join_processes(fn)
> Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 395, in _join_processes
> Nov 11 03:04:20     self._check_return_codes(elapsed_time)
> Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 431, in _check_return_codes
> Nov 11 03:04:20     raise RuntimeError(error)
> Nov 11 03:04:20 RuntimeError: Processes 0 1 exited with error code 10
> Nov 11 03:04:20 
> Nov 11 03:04:20 ======================================================================
> Nov 11 03:04:20 ERROR [1.920s]: test_reduce_sum_cuda (__main__.TestDistBackendWithFork)
> Nov 11 03:04:20 ----------------------------------------------------------------------
> Nov 11 03:04:20 Traceback (most recent call last):
> Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 278, in wrapper
> Nov 11 03:04:20     self._join_processes(fn)
> Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 395, in _join_processes
> Nov 11 03:04:20     self._check_return_codes(elapsed_time)
> Nov 11 03:04:20   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 431, in _check_return_codes
> Nov 11 03:04:20     raise RuntimeError(error)
> Nov 11 03:04:20 RuntimeError: Processes 0 1 exited with error code 10
> ```

yes just reverted it and will investigate and submit it again.
